workers/datasets_based/src/datasets_based/worker_factory.py

```diff
@@ -17,34 +21,68 @@
 from datasets_based.workers.splits import SplitsWorker
 
 
-class DatasetBasedWorkerFactory(WorkerFactory):
-    def __init__(self, app_config: AppConfig) -> None:
+class BaseWorkerFactory(ABC):
+    """
+    Base class for worker factories. A worker factory is a class that creates a worker.
+
+    It cannot be instantiated directly, but must be subclassed.
+    """
+
+    def create_worker(self, job_info: JobInfo) -> Worker:
+        return self._create_worker(job_info=job_info)
+
+    @abstractmethod
+    def _create_worker(self, job_info: JobInfo) -> Worker:
+        pass
+
+
+class WorkerFactory(BaseWorkerFactory):
+    def __init__(self, app_config: AppConfig, processing_graph: ProcessingGraph) -> None:
         self.app_config = app_config
+        self.processing_graph = processing_graph
```

Comment:
There is one thing I don't understand: the `processing_graph` continues being part of the `app_config`? So you are passing it here twice (separate and within the app_config)?

Reply:
I mean, if you are trying to decouple them, then `ProcessingGraph` shouldn't be removed from `AppConfig` (in workers/datasets_based/src/datasets_based/config.py)?

Reply:
oh, you're right.

This PR is a part I extracted from a bigger PR where I'm moving processing_graph outside of the app_config and I forgot to remove it. I'll fix this for now, for the sake of coherence